### PR TITLE
Get rid of the need for "with" translation

### PR DESCRIPTION
**Pull request by lexmag**

### CODE REVIEW

File: lib/elixir/src/elixir_exp.erl [hunk 1]
```diff
@@ -236,17 +236,13 @@ expand({'cond', Meta, [KV]}, E) ->
...
-  {EClauses, EC} = elixir_exp_clauses:'case'(Meta, KV, EE),
-  FClauses =
-    case (lists:keyfind(optimize_boolean, 1, Meta) == {optimize_boolean, true}) and
-         elixir_utils:returns_boolean(EExpr) of
-      true  -> rewrite_case_clauses(EClauses);
-      false -> EClauses
-    end,
-  {{'case', Meta, [EExpr, FClauses]}, EC};
+expand({'case', Meta, [Expr, KV]}, Env) ->
+  case lists:keyfind(export_vars, 1, Meta) of
```
Comment (josevalim) [inline lib/elixir/src/elixir_exp.erl]: Maybe use `proplists:get_value/3`?

File: lib/elixir/test/elixir/kernel/with_test.exs [hunk 2]
```diff
@@ -78,7 +78,7 @@ defmodule Kernel.WithTest do
 
   test "invalid else option" do
     assert_raise CompileError, "nofile:1: expected -> clauses for else in with",  fn ->
```
Comment (josevalim) [inline lib/elixir/test/elixir/kernel/with_test.exs]: Let's move any compile error from this file to expansions_test.

File: lib/elixir/src/elixir_translator.erl [hunk 3]
```diff
@@ -248,6 +250,15 @@ translate(Other, S) ->
 
 %% Helpers
 
+translate_case(true, Meta, Expr, KV, S) ->
```
Comment (josevalim) [inline lib/elixir/src/elixir_translator.erl]: :+1:

File: lib/elixir/src/elixir_translator.erl [hunk 4]
```diff
@@ -83,10 +83,12 @@ translate({'cond', CondMeta, [[{do, Pairs}]]}, S) ->
 %% Case
 
 translate({'case', Meta, [Expr, KV]}, S) ->
-  Clauses = elixir_clauses:get_pairs(do, KV, match),
-  {TExpr, NS} = translate(Expr, S),
-  {TClauses, TS} = elixir_clauses:clauses(Meta, Clauses, NS#elixir_scope{extra=nil}),
-  {{'case', ?ann(Meta), TExpr, TClauses}, TS#elixir_scope{extra=NS#elixir_scope.extra}};
+  case lists:keyfind(export_vars, 1, Meta) of
```
Comment (josevalim) [inline lib/elixir/src/elixir_translator.erl]: proplists here too.

Comment (lexmag) [inline lib/elixir/src/elixir_translator.erl]: `lists` use made sense before, at some early step, when I used `keytake`.

File: lib/elixir/src/elixir_exp.erl [hunk 5]
```diff
@@ -428,6 +420,21 @@ expand(Other, E) ->
 
 %% Helpers
 
+expand_case(true, Meta, Expr, KV, E) ->
+  assert_no_match_or_guard_scope(Meta, 'case', E),
+  {EExpr, EE} = expand(Expr, E),
+  {EKV, EK} = elixir_exp_clauses:'case'(Meta, KV, EE),
+  RKV =
+    case proplists:get_value(optimize_boolean, Meta, false) and
```
Comment (michalmuskala) [inline lib/elixir/src/elixir_exp.erl]: wouldn't `andalso` be better here?

Comment (lexmag) [inline lib/elixir/src/elixir_exp.erl]: Yup, will change.

**Pull request merged**